### ISSUE TITLE
Kanban + custom group-by breaks all visibility and read-only restrictions

### ISSUE THREAD

**Issue by elgow** [odoo/odoo#19390]
Impacted versions: 9, 10, 11

Steps to reproduce: 
1) Define a kanban view that hides some fields and marks them as read-only, e.g. 
`<field name="my_field" invisible="True" readonly="True"/>`
2) Use Group By to show a custom grouping by one of those fields in the kanban view.
3) Drag/drop kanban records between columns to alter values in supposedly read-only fields that should not even be visible.

Current behavior:  No  ability to prevent users from seeing and modifying fields that they shouldn't in a kanban view with group-by.

Expected behavior:  As with other views, allow the view to specify fields that should be invisible and/or read-only, and properly enforce those restrictions in the group-by drop-down and column drag/drop.

Video/Screenshot link (optional):

**Comment by elgow** [comment]
Does Odoo 11 fix this severe design flaw in the kanban view, or is it still a free pass to modify any field?

**Comment by elgow** [comment]
Just tried it out on Odoo 11. Bad news. It seems that though they extensively re-wrote the kanban view they didn't bother to fix this glaring problem. You can still use group-by + kanban view to change the value of pretty much any model field, whether you should be able to see it or not. Wow.

**Comment by gdgellatly** [comment]
@ged-odoo can you take a look at this please.  I don't agree with the original report, but still moving actual real readonly records on kanban shouldn't happen.  The best way I found to replicate is

On Runbot
- Go to Sales
- Group By Customer on Kanban.
- Cool I can move quotes to different customers (nice, I like it).
- Not so cool, I can move confirmed sales to different customers (although the customer doesn't change).

@elgow I don't think invisible does what you think it does.  Usually invisible is used in a tree or kanban precisely because we want to use it in a groupby.  There are hundreds of examples of desirable groupbys with invisible fields.  Invisible doesn't define a users access, or even ability to write, for those you want `groups` attributes and/or acls and record rules.  

Its a bit the same with your readonly definition, you aren't defining security there, e.g. a users write access, but rather just the behaviour of that single element on that page at that location.  Again using acls/record rules/groups is the correct way.

Still, actual real readonly fields - like a customer on a confirmed sale order shouldn't be able to be moved, even if it doesn't write.

**Comment by Yenthe666** [comment]
@gdgellatly actually, moving the record seems to update the customer too. Anyways I don't really like that you can move an already readonly record, that shouldn't really be the case. 😄 
I guess (@)ged-odoo will give us feedback.

**Comment by elgow** [comment]
@gdgellatly I can see your point about invisible=True not necessarily ruling out groupby, but the notion of readonly=True in a view is precisely that the field should not be editable from that view. I don't see why kanban should be an exception to the need for that. In particular, in kanban you can only set the value in a field to a value held by another record. A field for a personal identifier may need to be editable from a form view but, for safety, be forbidden from being set to another record's value in a kanban. 

As for invisible, I think it's undeniable that some fields may exist that should not be visible to users, even if they can be set in the background for those users. To their credit, Odoo at least does not present the ID field for grouping, which would be absurd. I have not tested this, but I expect that they don't provide the same feature for other fields that have SQL unique constraints that make them unsuitable for grouping.  When records are grouped by the create and write audit fields are moved they silently don't actually write the new value to the DB. That makes the behavior strange, but at least not destructive of the audit function. 

My opinion is still that it should be up to the module developer to decide what fields the user can see, group by, and edit from the kanban view. As it stands, the developer has very little control of it.

**Comment by ged-odoo** [comment]
@elgow @gdgellatly This is definitely unwanted behaviour (talking about the readonly thing).  When we rewrote the views, our priority was to reproduce the previous semantics, with test.  Obviously, when we found out that the previous behaviour was stupid/insane/wrong, we did the proper change.  I just think that we did not notice that the kanban view allow writing on readonly fields.  We definitely did not intend that.

This will be fixed in saas-16 (and forward ported to master).

**Comment by elgow** [comment]
@ged-odoo Please see my earlier comment explaining why a field that is not read-only for a given user/group may want to be read-only in a kanban view. I can easily make a field writable from one form and not from another, no? Why should kanban be different, especially when the writes that it can do always copy a value from another record? I take it that you agree that there is value in being able to declare a field to be read-only in a specific view, or you would not have built that capability into all your other views. 

BTW, this is made more urgent by the fact that (in 9.0 and 10.0 anyway) there is no simple way to turn off the group-by button. This means that the group-by+kanban combo produces an unmanageable access point where users can do things that module designers might need to prevent them from doing. No matter how you rationalize it, this is bad.

**Comment by elgow** [comment]
@ged-odoo @gdgellatly Actually, because kanban does not follow the usual edit/save/discard pattern it's far more important to be able to make fields read-only in kanban than in other views. Once a card is dragged and dropped in kanban the DB write takes place immediately and there is, AFAIK, no undo button. I work in a hospital. Imagine if someone drags a kanban card for a patient and changes a test or a diagnosis. How do they fix what they've done? 

Clearly your suggestion that I rely on the group access or ACLs to prevent it is not realistic, as those fields need to be writable by my users in the proper context.  What I need is the ability to prevent accidents in the kanban view by limiting that view's ability to write fields other than the very few where it's appropriate. This rogue ability of kanban to instantly alter fields with no confirmation step is a potential safety issue in my application, so I hope you will not stick to the "that's as designed" response. It's clearly a problematic design.

**Comment by ged-odoo** [comment]
@elgow what are you talking about? noone said "that's as designed".  Also, @gdgellatly is totally correct, you should not rely on the kanban view to enforce a business requirement.  Even when we will fix this one, nothing prevents someone from doing a rpc to write on a record with a little javascript.  The only proper security needs to be done by the server.  Sometimes with ACL/groups, sometimes with an override of the write method.

If you have a record that should not be updateable in some case, then you need to prevent that in python.

**Comment by elgow** [comment]
My apologies, I responded to you from @gdgellatly's response of 
> Its a bit the same with your readonly definition, you aren't defining security there, e.g. a users write access, but rather just the behaviour of that single element on that page at that location. Again using acls/record rules/groups is the correct way.

You responded that you will be fixing the move of read-only model fields, and I do appreciate that. It does not, however, solve the problem.

As I detailed in my comments, I see the problem in the way that the kanban view permits unsophisticated users, who aren't going to make an rpc call from code, to make changes that they should not through a simple drag and drop for which there is no confirmation step and no undo feature. The kanban view makes it impossible to hide fields from this sort of writing, and very difficult to even hide the entire group-by feature to prevent it. 

In short, the kanban needs the features for hiding fields that are present in the other views, and it needs them even more because of the one-click immediate write with no discard or undo. This design is not safe.

**Comment by elgow** [comment]
@ged-odoo To your point "If you have a record that should not be updateable in some case, then you need to prevent that in python.", can you elaborate a bit? If I have a field that should be writable by a user from a form view but not from a kanban view, how can I tell on the server side which view is trying to write to the field? That would be a requirement if I'm to follow your advice.

**Comment by gdgellatly** [comment]
Setting a context flag would do it easily enough.  Hundreds of example of that in the code although not for this specific case.  Usually its for calculating computationally expensive fields that are of no point or not creating secondary records when you have 2 create functions for an inherited object.

**Comment by elgow** [comment]
@gdgellatly I'm sorry for being obtuse here, but there's one thing I don't understand. Since the kanban view with group-by exposes all model fields w/o my having referenced them in the view, where would I set the context flag? If a user does group-by on a field that I've not named in the view and then drag/drop writes to that field wouldn't the write include only that field, an not other fields with any context that may be set on them?  Sorry but I'm just missing the means to tell the back end write() method which view is writing.  

Even if it's possible to do that, I'd still suggest that making a view where by default every field is visible and more easily writable than in the views that require edit/save is a very dubious design choice which can pose a safety hazzard.

**Comment by elgow** [comment]
@gdgellatly @ged-odoo I'd like to suggest a solution to what ails the kanban view. It has two parts:

1) Make the readonly="True" setting on a field work as one would expect it to based on its behavior in other view types. 

2) Provide an explicit means to specify the fields that will appear in the group-by drop down. This could be modeled on the search view, as that provides a very nice, intuitive means of controlling what the user search experience looks like. 

With these two things the kanban  view would be both safer and more usable. The user would not be faced with a group-by dropdown that contains fields that they've never heard of and don't understand. If it works for search, why not for group-by?

**Comment by gdgellatly** [comment]
@elgow  You could do your second point in under a minute yourself just by overriding the search view template and remove/comment the add custom group href.  Then you would only have explicit groupby's in the UI.

**Comment by elgow** [comment]
@gdgellatly Wow. Thank you. That is fantastic. I need to study the search view template and see how this will work, but if it's that simple then my problem is mostly solved. I can eliminate most fields from the group-by to avoid any unwanted writes. Depending on how the move of read-only fields issue gets solved this may be a 100% fix of every concern I've expressed in this issue. 

Thanks again.

**Comment by gdgellatly** [comment]
@elgow Note the could rather than should, beware the unintended consequences.  Personally its something I'd never consider doing, just because I don't believe I can anticipate every user requirement, but I suppose if it's one tightly controlled site it might be OK.  A proper implementation would probably be to override the Groupable widget of search views (sorry I forget the name, I'm at a Windows PC so just going by memory of source) but would take a lot more development and thinking.

**Comment by elgow** [comment]
@gdgellatly I can identify a few fields that I don't want exposed to the user for grouping, but as I looked into this I realized that the grouping is not really the problem. Grouping in the tree/list view is a very nice feature that poses no problem because it doesn't update the group-by field. The real problem with kanban is the unconstrained writing to fields with no edit/save process and no undo.  I'm hoping that when Odoo addresses this bug they will do it properly rather than just making the kanban refuse to drag/drop fields that are read-only at the model level. 

What's needed is to provide in the kanban view what the other views already have, a means to tell the view that a field is read-only in that view. Having to rely on making fields truly read-only at the model level is not the answer. If it were then the other views would not have a read-only setting.

**Comment by mge-odoo** [comment]
Hi,

Thank you for reporting the bug and for your implication in Odoo.

The issue has been fixed in saas-16 in the following revision: https://github.com/odoo/odoo/commit/5ba6d93810ba447f94ca2bb0d6804865a32b5cbf

The `readonly` attribute on the view is now also taken into account.
A small improvement has also been introduced ; instead of completely preventing the drag&drop (the resequence inside a column should still be possible even if the grouped field is readonly), we use the `containment` attribute to prevent the record from being dropped outside of its column.

It will soon be forward-ported in 11.0

Regards,

Martin

**Comment by elgow** [comment]
Thank you Martin for the fix and further improvements. This will let me make better use of the kanban view to provide features for my users. 

  Ed
